lib/Trash/PageTrashBackend.php

```diff
@@ -0,0 +1,694 @@
+		if (!($item instanceof CollectivePageTrashItem)) {
+			throw new \LogicException('Trying to remove normal trash item in collective trash backend');
+		}
+		$user = $item->getUser();
+		[, $collectiveId] = explode('/', $item->getTrashPath());
+		$node = $this->getNodeForTrashItem($user, $item);
+		if ($node === null) {
+			throw new NotFoundException();
+		}
+
+		// Get original parent folder of item to revert subfolders further down
+		$targetFolder = $this->collectiveFolderManager->getFolder((int)$collectiveId);
+		$targetFolderPath = substr($item->getOriginalLocation(), 0, -strlen($item->getName()));
+		if ($targetFolderPath) {
+			try {
+				$targetFolder = $targetFolder->get($targetFolderPath);
+			} catch (NotFoundException $e) {
+				$targetFolder = null;
+			}
+		}
+
+		// Get pageId for deleting page from collective page database
+		$deletePageId = null;
+		if ($node instanceof Folder) {
+			// Try to use index page if folder is deleted
+			if (null !== $indexNode = $node->get(PageInfo::INDEX_PAGE_TITLE . PageInfo::SUFFIX)) {
+				$deletePageId = $indexNode->getId();
+			}
+		} else {
+			$deletePageId = $node->getId();
+		}
+
+		if ($node->getStorage()->unlink($node->getInternalPath()) === false) {
+			throw new \Exception('Failed to remove item from trashbin');
+		}
+
+		$node->getStorage()->getCache()->remove($node->getInternalPath());
+		if ($item->isRootItem()) {
+			$this->trashManager->removeItem((int)$collectiveId, $item->getName(), $item->getDeletedTime());
+		}
+
+		if (!is_null($this->versionsBackend)) {
+			$this->versionsBackend->deleteAllVersionsForFile($collectiveId, $item->getId());
+		}
+
+		// Delete page from collective page database
+		if ($deletePageId) {
+			$this->pageMapper->deleteByFileId($deletePageId);
+		}
+
+		// Also remove attachments folder if it exists
+		if (null !== $attachmentsFolderItem = $this->findAttachmentFolderItem($user, $collectiveId, $item)) {
+			$this->removeItem($attachmentsFolderItem);
+		}
+
+		// Try to revert subfolders of target folder parent
+		if ($targetFolder) {
+			try {
+				NodeHelper::revertSubFolders($targetFolder->getParent());
+			} catch (\OCA\Collectives\Service\NotFoundException | \OCA\Collectives\Service\NotPermittedException $e) {
+			}
+		}
+	}
+
+	/**
+	 * @param IUser                   $user
+	 * @param int                     $collectiveId
+	 * @param CollectivePageTrashItem $item
+	 *
+	 * @return TrashItem|null
+	 * @throws NotPermittedException
+	 */
+	private function findAttachmentFolderItem(IUser $user, int $collectiveId, CollectivePageTrashItem $item): ?TrashItem {
+		$attachmentsPrefix = ".attachments.";
+		if (strpos($item->getName(), $attachmentsPrefix) === 0) {
+			// Passed item is already an attachment folder
+			return null;
+		}
+
+		$trashFolder = $this->getTrashFolder($collectiveId);
+		$deletedTime = $item->getDeletedTime();
+		// Search for attachments folder with deleted time up to two seconds after item deleted time
```

Comment:
Well, the attachments are deleted by a listener from the Text app, see https://github.com/nextcloud/text/blob/5d1095c5d8c91175f21d5c09dc84a7f6701f1b44/lib/Listeners/BeforeNodeDeletedListener.php#L49. My thought was that theoretically this could happen with a short delay, at least after the switch to next second. That's why I implemented it that way. But maybe I'm overthinking this and there's no practical scenario when the deleted timestamps from the markdown file and its attachments folder could vary? :thinking:

Reply:
That's a part where i was not absolutely sure about the flow, I thought file and folders are deleted in the same routine. Then it might have been possible to enforce the same timestamp, be there a second flip or not. Having a different component involved changes things. Also, deleting a lot on a slow,m perhaps external, storage during heavy times might slow down operations.